### PR TITLE
Lde/pcs in verifier

### PR DESCRIPTION
**Pull request by ledwards2225**
# Description

Adding Gemini/Shplonk/KZG + final pairing check functionality to the Honk verifier.

# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [x] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [x] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [x] There are no circuit changes, OR a cryptographer has been assigned for review.
- [x] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [x] The branch has been rebased against the head of its merge target.
- [x] I'm happy for the PR to be merged at the reviewer's next convenience.
- [x] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [x] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

File: cpp/src/aztec/honk/composer/composer_helper/composer_helper.cpp [hunk 1]
```diff
@@ -103,8 +103,11 @@ std::shared_ptr<waffle::verification_key> ComposerHelper<CircuitConstructor>::co
             auto selector_poly_commitment = commitment_key.commit({ selector_poly_coefficients, proving_key->n });
             circuit_verification_key->constraint_selectors.insert(
                 { selector_commitment_label, selector_poly_commitment });
-
-        } else if (selector_poly_info.source == waffle::PolynomialSource::PERMUTATION) {
+            // TODO(luke): Adding commitments to polys with label OTHER to 'permutation_selectors' for now. In the
```
Comment (codygunton) [inline cpp/src/aztec/honk/composer/composer_helper/composer_helper.cpp:106]: As we discussed, this is a temporary measure. We will aim for more uniform handling of the polynomials.

File: cpp/src/aztec/honk/composer/standard_honk_composer.test.cpp [hunk 2]
```diff
@@ -297,10 +297,11 @@ TEST(standard_honk_composer, test_verification_key_creation)
...
     auto verification_key = composer.compute_verification_key();
     // There is nothing we can really check apart from the fact that constraint selectors and permutation selectors were
-    // committed to, we simply check that the verification key now containset he appropriate number of constraint and
-    // permutation selector commitments. This method should work with any future arithemtization
+    // committed to, we simply check that the verification key now contains the appropriate number of constraint and
+    // permutation selector commitments. This method should work with any future arithemtization.
+    // Note(luke): permutation_selectors contains 2*program_width perm/id commitments + 2 lagrange poly commitments.
     EXPECT_EQ(verification_key->constraint_selectors.size(), composer.circuit_constructor.selectors.size());
-    EXPECT_EQ(verification_key->permutation_selectors.size(), composer.program_width * 2);
+    EXPECT_EQ(verification_key->permutation_selectors.size(), composer.program_width * 2 + 2);
```
Comment (codygunton) [inline cpp/src/aztec/honk/composer/standard_honk_composer.test.cpp:304]: Adding the Lagrange polynomials.

File: cpp/src/aztec/honk/pcs/kzg/kzg.test.cpp [hunk 3]
```diff
@@ -41,4 +43,106 @@ TYPED_TEST(BilinearAccumulationTest, single)
     this->verify_accumulators(acc, result_acc);
 }
 
+/**
```
Comment (codygunton) [inline cpp/src/aztec/honk/pcs/kzg/kzg.test.cpp:46]: Thanks for adding this test.

File: cpp/src/aztec/honk/proof_system/verifier.cpp [hunk 4]
```diff
@@ -63,6 +66,7 @@ template <typename program_settings> Verifier<program_settings>& Verifier<progra
         id(X)
 
     Univariate evaluations:
+        a_0_pos = Fold_{r}^(0)(r),
```
Comment (codygunton) [inline cpp/src/aztec/honk/proof_system/verifier.cpp]: This should go away.

File: cpp/src/aztec/honk/proof_system/verifier.cpp [hunk 5]
```diff
@@ -72,6 +76,8 @@ template <typename program_settings> Verifier<program_settings>& Verifier<progra
     Commitments:
         [w_i]_1,        i = 1,2,3
         [z_perm]_1,
+        [Fold_{r}^(0)]_1,
```
Comment (codygunton) [inline cpp/src/aztec/honk/proof_system/verifier.cpp]: This should go away.

**Review by codygunton — CHANGES_REQUESTED** [review]
We discussed on a call and I made some small notes recording our discussion. Looks good pending those.

File: cpp/src/aztec/honk/proof_system/verifier.cpp (hunk 5 above)
Comment (ledwards2225) [reply cpp/src/aztec/honk/proof_system/verifier.cpp]: Done

File: cpp/src/aztec/honk/proof_system/verifier.cpp (hunk 4 above)
Comment (ledwards2225) [reply cpp/src/aztec/honk/proof_system/verifier.cpp]: Done

**Review by codygunton — APPROVED** [review]

**Pull request merged**